libs/src/services/creatorNode/CreatorNodeSelection.js

```diff
@@ -0,0 +1,153 @@
+        isBehind: true,
+        isConfigured: false,
+        geographicData: {}
+        // todo: add other criteria we should evaluate like diskUsage, currentNumberOfRequests
+      }
+    })
+
+    const healthyEndpoints = []
+    const healthyServices = []
+    for (const service of services) {
+      const start = Date.now()
+      try {
+        // Check health and sync status, and then update stats per service
+        let healthCheckResp = await axios({
+          method: 'get',
+          url: map[service].healthCheckEndpoint
+        })
+        healthCheckResp = healthCheckResp.data.data
+
+        map[service].healthCheckResponseTime = Date.now() - start
+        map[service].version = healthCheckResp.version
+        const { isBehind, isConfigured } = await this.creatorNode.getSyncStatus(service)
+        map[service].isBehind = isBehind
+        map[service].isConfigured = isConfigured
+        map[service].geographicData = {
+          country: healthCheckResp.country,
+          latitude: healthCheckResp.latitude,
+          longitude: healthCheckResp.longitude
+        }
+
+        // If service is not configured, behind in blocks, or its major + minor version is behind, mark as unhealthy
+        map[service].isHealthy = isConfigured && !isBehind &&
+          this.ethContracts.hasSameMajorAndMinorVersion(this.currentVersion, healthCheckResp.version)
+
+        // Add to healthyServices list; used and sorted to select primary and secondaries
+        if (map[service].isHealthy) {
+          healthyEndpoints.push(service)
+          healthyServices.push({ endpoint: service, version: map[service].version, responseTime: map[service].healthCheckResponseTime })
+        }
+      } catch (e) {
+        console.error(`CreatorNodeSelection - Error with checking ${service} health: ${e}`)
+      }
+    }
+    this.decisionTree.push({ stage: 'Filtered Out Known Unhealthy', val: healthyEndpoints })
+
+    // Prioritize higher versions for selected creator nodes
+    this.sortBySemver(healthyServices)
+
+    // Index 0 of healthyServices will be the most optimal creator node candidate
+    const primary = healthyServices[0] ? healthyServices[0].endpoint : null
+
+    // Index 1 to n of healthyServices will be sorted in highest version -> lowest version
+    // Return up to numberOfNodes-1 of secondaries that is not null and not the primary
+    let numSecondariesReturned = 0
+    const secondaries = healthyServices
+      .filter(service =>
+        service && service.endpoint !== primary && numSecondariesReturned++ < this.numberOfNodes
+      )
+      .map(service => service.endpoint)
+
+    this.decisionTree.push({ stage: 'Made A Selection of a Primary and Secondaries', val: { primary, secondaries: secondaries.toString() } })
+
+    console.info('CreatorNodeSelection - final decision tree state', this.decisionTree)
+
+    // not sure about the structure of services
+    return { primary, secondaries, services: map }
```

Comment:
the structure of `services` (correlated with /version response) now contains /health_check response. I added the longitude, latitude, and country to the health_check endpoint. so this new services will basically have more data than the old services

Reply:
the old service has this typing so with the other /health_check data it might break
https://github.com/AudiusProject/audius-client/blob/f52a163a695e2bc9df93859db4e86c52f342779c/src/containers/service-selection/store/slice.ts#L4

Reply:
discussed offline -- typescript is impactful at compile time and not runtime, so my change in service structure will not impact usage